cmd/skopeo/sync.go

```diff
@@ -0,0 +1,594 @@
+		}
+		sourceReferences = append(sourceReferences, ref)
+	}
+	return sourceReferences, retErr
+}
+
+// Given an image reference as a local directory, returns all the image
+// references available at the given path.
+func imagesToCopyFromDir(dirPath string) (sourceReferences []types.ImageReference, retErr error) {
+
+	if _, err := os.Stat(dirPath); err != nil {
+		return []types.ImageReference{},
+			errors.WithMessage(err, fmt.Sprintf("Error checking for images in source path %q", dirPath))
+	}
+
+	err := filepath.Walk(dirPath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if !info.IsDir() && info.Name() == "manifest.json" {
+			ref, err := getImageReference(directory.Transport, fmt.Sprintf("%s", filepath.Dir(path)))
+			if err != nil {
+				return errors.WithMessage(err, fmt.Sprintf("Error while creating image referenced for path %s",
+					filepath.Dir(path)))
+			}
+			sourceReferences = append(sourceReferences, ref)
+			return filepath.SkipDir
+		}
+		return nil
+	})
+
+	if err != nil {
+		return []types.ImageReference{},
+			errors.WithMessage(err, fmt.Sprintf("Error walking the path %q", dirPath))
+	}
+
+	return
+}
+
+// Given a source URL and context, returns a list of tagged image references to
+// be used as sync source.
+func syncFromURL(sourceURL *url.URL, sourceCtx *types.SystemContext) (repoDescriptor, error) {
+	repoDesc := repoDescriptor{
+		Context: sourceCtx,
+	}
+
+	switch transports.Get(sourceURL.Scheme) {
+	case docker.Transport:
+		srcRef, err := getImageReference(docker.Transport, fmt.Sprintf("//%s%s", sourceURL.Host, sourceURL.Path))
+		if err != nil {
+			return repoDesc, errors.WithMessage(err, "Error while parsing destination")
+		}
+
+		imageTagged, err := isTagSpecified(sourceURL.Host + sourceURL.Path)
+		if err != nil {
+			return repoDesc, err
+		}
+		if imageTagged {
+			repoDesc.TaggedImages = append(repoDesc.TaggedImages, srcRef)
+			break
+		}
+
+		repoName := fmt.Sprintf("//%s", path.Join(sourceURL.Host, sourceURL.Path))
+		repoDesc.TaggedImages, err = imagesToCopyFromRegistry(srcRef, repoName, sourceCtx)
+		if err != nil {
+			return repoDesc, err
+		}
+	case directory.Transport:
+		dirPath, err := dirPathFromURL(sourceURL)
+		if err != nil {
+			return repoDesc, errors.WithMessage(err, "Error processing source URL")
+		}
+
+		repoDesc.DirBasePath = dirPath
+		repoDesc.TaggedImages, err = imagesToCopyFromDir(dirPath)
+		if err != nil {
+			return repoDesc, err
+		}
+	}
+
+	if len(repoDesc.TaggedImages) == 0 {
+		return repoDesc, errors.New("No images to sync found in SOURCE")
+	}
+
+	return repoDesc, nil
+}
+
+// Given a yaml file and a source context, returns a list of repository descriptors,
+// each containing a list of tagged image references, to be used as sync source.
+func syncFromYaml(yamlFile string, sourceCtx *types.SystemContext) (repoDescList []repoDescriptor, err error) {
```

Comment:
I’m still unsure about this — what does the `.yaml` format do that couldn’t just as well be done with an ordinary shell script calling `skopeo`? Allow recording credentials in a file so that they don’t show up in **ps**(1)?

Reply:
The use case is the ability to specify multiple images (and any tags combination) hosted on multiple registries (each one with its own auth method) in a yaml file. 

You then call `sync` one time only with:
`skopeo sync --source yaml images.yaml dir:syncedImages`

Instead of:
```
skopeo sync [opts] docker://registry1/image1 dir:destDir
skopeo sync [opts] docker://registry1/image2 dir:destDir
...
skopeo sync [opts] docker://registry1/imageN dir:destDir
skopeo sync [opts] docker://registry2/image1 dir:destDir
...
skopeo sync [opts] docker://registryN/imageN dir:destDir
```

Reply:
> The use case is the ability to specify multiple images (and any tags combination) hosted on multiple registries (each one with its own auth method) in a yaml file.

Sure, the config file allows doing that, but in what way is it materially different from writing a shell script just like your second example? Or, really,

```sh
for source in \
    docker://registry1/image1 \
    docker://registry1/image2 \
...
    docker://registry1/imageN \
    docker://registry2/image1 \
...
    docker://registryN/imageN \
    ; do
        skopeo sync [opts] "$source" dir:destDir
done
```

(I’m not saying the `yaml` config definitely does not make sense, just that it’s not obviously clear to me that it is worth defining and maintaining another stable file format.)

Reply:
> Sure, the config file allows doing that, but in what way is it materially different from writing a shell script just like your second example?

One advantage is that yaml can be generated from some other tool using common libraries, while the same is not so obvious for bash. 

E.g. we have an [Helm plugin to mirror Chart repositories](https://github.com/openSUSE/helm-mirror#helm-mirror-plugin). It generates yaml files that can be consumed by `skopeo sync`, so that also the images specified in the charts can be mirrored.

We thought it's more convenient to do it in this way, instead of having another tool just to read the list of images/ registers configuration and run `skopeo sync` from it.